### ISSUE TITLE
Consider using the streaming API for parsing CSV (> v0.4.0)

### ISSUE THREAD

**Issue by emmenko** [commercetools/sphere-node-product-csv-sync#56]
http://csv.adaltas.com/parse/
http://csv.adaltas.com/parse/examples/

**Comment by nkuehn** [comment]
I am currently not able to import my large (icecat) dataset because the importer crashes node low-level due to the file size.  There is no output but the following. 
```
#
# Fatal error in ../deps/v8/src/handles.h, line 48
# CHECK(location_ != NULL) failed
#
==== C stack trace ===============================
 1: ??
 2: ??
 3: ??
 4: ??
 5: ??
 6: ??
 7: ??
Illegal instruction: 4
```
it is clearly related to the CSV file size (I tried a subset of the file and it works). 

I am not sure whether it is necessary to switch to stream reading all through the codebase, we should first do the low hanging fruit like not representing the file twice in memory (currently the raw file content is put in a variable which is then passed to the validator, which again holds a representation -> https://github.com/sphereio/sphere-node-product-csv-sync/blob/master/src/coffee/import.coffee#L50 ).  There is a comment in the validator class already for this issue.

**Comment by nkuehn** [comment]
P.S.:  the file that causes the crash is 257 MB large, which is not completely atypical for bigger data feeds in ecommerce.

**Comment by hajoeichler** [comment]
The rewrite idea is mentioned some lines above: https://github.com/sphereio/sphere-node-product-csv-sync/blob/master/src/coffee/import.coffee#L33

It's in progress - but far from being ready.

**Comment by emmenko** [comment]
That's why we started with the [new node-cli streaming ](https://github.com/sphereio/sphere-node-cli) and [product import](https://github.com/sphereio/sphere-product-import) ;)

**Comment by nkuehn** [comment]
@hajoeichler nice, thank you!

Nevertheless: I'm not sure whether the CSV line chunking solves more problems than it creates.  Even if you can smartly detect product variant blocks belonging together, it would cement the problematic requirement that product variants need to be in direct sequence and well prepared (which practically means that you need to strap the csv-mapper in front of the import which loads everything into memory, too).  

TL;DR:  Someone needs to load the whole file anyways to do the variant grouping, so let's pay attention to not just moving the problem upstream into the preprocessing steps.

**Comment by nkuehn** [comment]
Status update: something seems to have improved in the last versions in node and the csv sync. I can now pass a higher old-gen heap parameter and the import runs through. 

I tried to free some of the (many) duplicate memory structures (raw CSV, parsed CSV, Product Objects that have already been imported), but it turned out to be not possible in a straightforward way because  of the deep promise nesting where pointers to data and structures not needed and more are still held somewhere (probably in the higher nesting levels). 

BTW: node showed a weird memory / GC behavior. The import of my test file runs through successfully from max Oldspace settings ca. 1.7GB upwards, but the more space I give it it will always take it and grow continuously during the actual API import part (both physical and virtual memory).
